adserver/views.py

```diff
@@ -751,6 +753,152 @@ def get_context_data(self, **kwargs):
         return context
 
 
+class AdvertiserGeoReportView(AdvertiserAccessMixin, GeoReportMixin, BaseReportView):
+
+    """A report for an advertiser broken down by geo."""
+
+    model = GeoImpression
+    template_name = "adserver/reports/advertiser-geo.html"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        start_date = context["start_date"]
+        end_date = context["end_date"]
+
+        country = self.request.GET.get("country", "")
+
+        advertiser_slug = kwargs.get("advertiser_slug", "")
+        advertiser = get_object_or_404(Advertiser, slug=advertiser_slug)
+
+        queryset = self.get_queryset(
+            advertiser=advertiser,
+            start_date=start_date,
+            end_date=end_date,
+            country=country,
+        )
+
+        report = AdvertiserGeoReport(
+            queryset,
+            # Index by date if filtering report to a single country
+            index="date" if country else None,
+            order="-date" if country else None,
+            max_results=None if country else self.LIMIT,
+        )
+        report.generate()
+
+        country_options = self.get_country_options(
+            self.get_queryset(
+                advertiser=advertiser,
+                start_date=context["start_date"],
+                end_date=context["end_date"],
+            )
+        )
+
+        context.update(
+            {
+                "advertiser": advertiser,
+                "report": report,
+                "country": country,
+                "country_options": country_options,
+                "country_name": self.get_country_name(country),
+                "export_url": "{url}?{params}".format(
+                    url=reverse("advertiser_geo_report_export", args=[advertiser.slug]),
+                    params=urllib.parse.urlencode(
+                        {
+                            "start_date": context["start_date"].date(),
+                            "end_date": context["end_date"].date()
+                            if context["end_date"]
+                            else "",
+                            "country": country or "",
+                        }
+                    ),
+                ),
+            }
+        )
+
+        return context
+
+
+class AdvertiserPublisherReportView(AdvertiserAccessMixin, BaseReportView):
+
+    """A report for an advertiser broken down by publishers where the advertisers ads are shown."""
+
+    template_name = "adserver/reports/advertiser-publisher.html"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        start_date = context["start_date"]
+        end_date = context["end_date"]
+
+        advertiser_slug = kwargs.get("advertiser_slug", "")
+        advertiser = get_object_or_404(Advertiser, slug=advertiser_slug)
+        report_publisher = Publisher.objects.filter(
+            slug=self.request.GET.get("publisher", "")
```

Comment:
We should probably change the GET arg on the publisher reports to just `advertiser`, instead of exposing `report_advertiser` in the URL heh.